accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+
+As such, there are several risks of adopting Sigstore for npm that are worth calling out:
+- No buy-in from the broader open source npm community.
+- Sigstore's public and free-to-use infrastructure is not able to meet our uptime requirements (>99.9%).
+- Sigstore is not able to secure the funding required to run the infrastructure with production-grade support, SLOs and on-call rotation.
+
+To mitigate some of these risks, GitHub is planning to work directly with Sigstore to define and support robust production-grade SLOs for uptime and reliability that meet the requirements for npm. This will involve being on-call and wearing the pager for production services.
+
+In addition, GitHub is planning to support Sigstore by:
+- Running the signed timestamp authority for Rekor as well as a Certificate Transparency monitor for Fulcio. This will help spread the trust outside the four walls of Sigstore.
+- Maintaining the Sigstore trust root and having a root key holder as part of the group of Sigstore root key holders.
+
+We also remain open to an alternative solution if Sigstore is not able to meet npm's requirements for uptime and support.
+
+## Detailed Explanation
+### Sigstore as a solution to package signing
+Package signing is a mitigation for package hijacking attacks. If maintainers sign their packages, end-users can verify that packages were produced by their corresponding owners. However, as of today, the main existing solution involves maintainer-managed PGP (or similar) keys, which has known [usability](https://www.usenix.org/legacy/events/sec99/full_papers/whitten/whitten.ps) [and](https://latacora.micro.blog/2019/07/16/the-pgp-problem.html) [security](https://blog.cryptographyengineering.com/2014/08/13/whats-matter-with-pgp/) [issues](https://www.wired.co.uk/article/efail-pgp-vulnerability-outlook-thunderbird-smime): registries that have implemented public key signing have seen either limited adoption or achieved adoption only by mandating it (e.g., Maven Central, which reports developer friction).
+
+[Sigstore](https://www.sigstore.dev/) is a promising alternative: it allows signing artifacts with short-lived “disposable/ephemeral keys” notarized against log-ins with [OpenID Connect](https://openid.net/connect/) (OIDC) identities (e.g., Google or GitHub accounts).
+
+Several package ecosystems have announced interest in adopting Sigstore as a way for maintainers to sign packages (e.g., PyPI, RubyGems, Maven Central). [OpenSSF](https://blog.chainguard.dev/sigstore-statement-for-openssf-call/) is also working to standardize on Sigstore for signing software.
+
+However, adopting something like Sigstore raises several major questions (beyond the risks detailed above):
+- What security benefits does Sigstore provide?
+- How should maintainers authenticate to sign packages? The solution should provide both high assurance in the authenticity of packages and respect maintainer privacy.
```

Comment:
There will inevitably be maintainers who object to signing for a variety of reasons. While I think it is reasonable to listen, a software repository also answers to the needs of end users.

For a maintainer a compromised package is perhaps embarrassing and distressing. For end users it is potentially catastrophic. And there are far more end users than maintainers. I don't see any other way to do the utilitarian calculus than to find that the needs of the end users for enhanced security are paramount.

Reply:
In any ecosystem where you have two cohorts of users, producers and consumers, it's simply not viable to do utilitarian calculus, because neither group can exist without the other. Both groups' needs must be weighed together.

eg, mandatory 2FA for maintainers is annoying and inconvenient, but the burden on maintainers is relatively small, and the benefits for maintainers and end users is exceedingly high, so it works out to enforce it.

If the burden is too large, or the benefits aren't bulletproof enough or high enough, it wouldn't likely be worth the tradeoffs.

Reply:
> Both groups' needs must be weighed together.

That is in fact the calculus. Two variables: cost to maintainers, cost to end users, integrated under a curve with respect to the spectrum of potential outcomes. The area under the curve is dominated by the impact on users because there are so many and because they are more negatively affected than maintainers by a compromise.

>If the burden is too large, or the benefits aren't bulletproof enough or high enough, it wouldn't likely be worth the tradeoffs.

The burden here is quite minimal. One click for the common case.

Reply:
I think it needs to be minimal for the uncommon cases too. Prolific maintainers can be eccentric, myself included, and doing both unpaid technical and emotional work does not tend to predispose humans to being charitable.

Reply:
> For a maintainer a compromised package is perhaps embarrassing and distressing.

Sure, but this system doesn't meaningfully prevent this from happening. It _does_ prevent a certain version of this from happening, which IMO isn't necessarily a compelling story for me to add it outside of end-users maybe wanting this... but end-users want a lot of things, many of which are unreasonable/invalid. Social pressure shouldn't be the primary motivator.

Reply:
> The burden here is quite minimal. One click for the common case.

Unfortunately, the most depended upon packages aren't the common case 😉

Reply:
Copying my answer from another comment. Part of the story is not placing any kind of maintenance overhead on maintainers, for example not having to manage keys and making the UX/DX seamless. Saying that, the real burden with this proposal is moving publishing to CI/CD.

Arguably publishing from CI/CD today is actually less secure than from your laptop where you can use 2FA. We can do better.

We're also working on a proposal which will allow authorising the publish using the same OIDC id token from the CI/CD system. So instead of using a long-lived access token to authorize the publish, you would log into your npm dashboard and set up which source repo is allowed to publish what package. The publish step then exchanges the OIDC id token for a short lived access token after the npm registry has checked the source repo from the id token matches the allowed one.

Once you've set up publishing in this way, you would not need to use long-lived access tokens and the risk of stolen credentials basically goes away if you also use 2FA on your npm account. The setup for this would be exactly the same as what we're proposing in this RFC so you essentially get source and build linking for free if you set up OIDC publishing. I imagine adoption would pick up a lot once we have both features in place and all major CI/CD providers are fully supported.

We also want to provide reusable workflows that easy the burden of moving publishing to ci/cd.

Reply:
>Prolific maintainers can be eccentric, myself included, and doing both unpaid technical and emotional work does not tend to predispose humans to being charitable.

I understand and this dynamic is familiar to me as an observer of MFA rollouts in RubyGems and PyPI. My general point is that package maintainers, while an important constituency, are not the _only_ constituency. Software repositories owe a responsibility to all their stakeholders. Sometimes their interests and preferences will be in conflict and one constituency will necessarily be privileged over another, even though this will cause distress to some members of other constituencies.

This is such a case. I don't revel in the added work for package maintainers, however slight it is. But I see it as unavoidably necessary, given the world we now live in.

Reply:
>Social pressure shouldn't be the primary motivator.

I don't think it is. This is about cold hard numbers. The second most common attack on supply chains is account takeover. The third most common is repository compromise. Signing is a countermeasure to reduce the risk of both of these and is complementary to other countermeasures like MFA and implementing TUF.

Reply:
We are getting requests from maintainers and consumers that they want tools to make more informed and safe choices about the dependencies they consume. The solution that we've landed on for this is signed build attestations from trusted environments.

As this RFC outlines simply running `publish` from a trusted build environment (e.g. actions) would be all that's needed to do this. The surface DX for customers is something we will work together on fine tuning, but it mostly boils down to "is there a flag or not".

Here is an [example workflow](https://github.com/MylesBorins/node-osc/blob/main/.github/workflows/create-release.yml) that both publishes to npm and makes a GitHub release, if the RFC moves in the direction I want to advocate for my project would "just start signing" if I had a version of the npm CLI that supported signing. IMHO that is a fairly low barrier to entry.

Longer term we are looking at ways to significantly streamline the publishing process from GitHub -> npm built on open technologies such as OIDC that could make the process for developers setting up automations a "click of a button" with no managed tokens. While this is the long term goal I think it is out of scope for this RFC.

Reply:
Why is my local machine not capable of being a trusted build environment? A staggering amount of npm's downloads already trust it implicitly and have for almost a decade.

Reply:
> Why is my local machine not capable of being a trusted build environment?

With this proposal we want to figure out a way to say: this package on the npm registry, it was built from this source and you can go view it at this link to audit it if you like, and have some guarantees that it wasn't falsified. There's no way afaik to verifiably link the source repository to the published package from your local machine. You could, as a maintainer, say your package is from repo x, branch y and commit z but there's no way to verify this is actually legit information.

Taking this information from the OIDC id token generated by the CI/CD environment means it's not possible to falsify the information without the CI/CD systems private key used to sign it. With a reusable workflow that controls the build setup, we can also go one step further and make sure the repo and branch that you're building from is the same one that's specified in the OIDC id token, making sure the build hasn't gone and checked out some other branch to the one the action was triggered on for example.

Reply:
@feelepxyz a CI/CD machine is just as attackable as my local machine, though - perhaps more so, since it's a known target that's exposed to the open web. Why can't my local machine - something that has proven to be more trustworthy so far than all CI providers I'm aware of, largely due to nobody trying to attack it I'm sure - be able to sign it?

Reply:
> @feelepxyz a CI/CD machine is just as attackable as my local machine, though - perhaps more so, since it's a known target that's exposed to the open web.

This is true and it definitely becomes a much more appealing target! 

> Why can't my local machine - something that has proven to be more trustworthy so far than all CI providers I'm aware of, largely due to nobody trying to attack it I'm sure - be able to sign it?

For me it isn't that I don't trust your personal build machine, its that we can't enforce any kind of provenance claims as to what repo the package was built from. An attacker that's publishing a malicious version of some package from their local machine could easily state that the package is from the original legitimate repo, without it actually reflecting the changes in the published package. 

What we could still do, and may want to enable this in future, is allow maintainers to sign from local machines using some identity that authenticates them as maintainers of the package, e.g. email based OIDC. This would then show *who* signed it but not *where* and *how*.

Reply:
> @feelepxyz a CI/CD machine is just as attackable as my local machine, though - perhaps more so, since it's a known target that's exposed to the open web. Why can't my local machine - something that has proven to be more trustworthy so far than all CI providers I'm aware of, largely due to nobody trying to attack it I'm sure - be able to sign it?

I would really argue that this isn't true, unless you're running something without any security controls. I would assume the CI/CD systems here would be follow standards, be pen tested, etc. Compare that with a local workstation that could be downloading arbitrary software, running arbitrary software in addition to the build you're running. That build could be attacked from a pretty unbounded number of vectors. Compare that with a CI/CD system that has adequate network, identity, etc. controls, there would be minimized attack surface.

> For me it isn't that I don't trust your personal build machine, its that we can't enforce any kind of provenance claims as to what repo the package was built from. An attacker that's publishing a malicious version of some package from their local machine could easily state that the package is from the original legitimate repo, without it actually reflecting the changes in the published package.

This doesn't fix the problem right? A CI/CD system that has been compromised can falsify some elements of the provenance just as well as a laptop.

Reply:
> This doesn't fix the problem right? A CI/CD system that has been compromised can falsify some elements of the provenance just as well as a laptop.

This is my point, I think - that there's nothing inherent about a CI/CD system that isn't also true about a personal laptop, and vice versa - the only difference is likelihoods due to competence, target awareness/attractiveness, which software is being run, etc.

Reply:
That's true. However, I think generally CI/CD is easier to secure compared to your laptop. CI/CD system can be audited, and configured through code, to ensure that for example that builds happen hermetically, only download pre-approved tools with a provenance trail. Your laptop unless it's managed by some other system doesn't haven't similar properties. Also in the case of packages that have multiple developers and maintainers it makes it more clear that packages are built through one system that has been approved by the project.

In the case of let's say GHA, you need to trust that GHA is doing the right things from a workflow perspective, OIDC, etc. GHA is taking on a lot of risk to their reputation, but most likely have resources to do so. If I build an npm package from my local laptop it's mostly on me. I could inadvertently have a rootkit on my machine, using a compromised OS kernel, have downloaded malicious software, etc. It is harder to trust anonymous actor's laptop compared to a potentially audited/threat modeled CI/CD system or service.

Reply:
Perhaps - but there are few personal laptop environments more objectively trustable than mine based on track record :-) without a doubt the majority will want to use this system, and probably should! However, if it's made a _requirement_ - explicitly, or implicitly via creating metadata that effectively punishes an author for not complying - that seems like it would be harming maintainers for a very dubious benefit.

Reply:
Trust has economies of scale. Asking me to trust every random laptop means that, to achieve a similar level of confidence to a highly defended central provider, I would need to interview a bunch of people who, as has been pointed out elsewhere in this discussion, might object to the added burden.

Put another way: I don't know you. I know Github. Given my current priors, I trust them more than you, and it's costly for both you and I to update those priors.